### DISCUSSION TITLE
forecasting - get fitted values

### DISCUSSION THREAD

**Discussion by eyalshafran** [alan-turing-institute/sktime#888] [Q&A] [answered]
When using statsmodel it is possible to access the fitted values after the model is fit. 
For example:
```
from sktime.datasets import load_airline
from statsmodels.tsa.api import ExponentialSmoothing as _ExponentialSmoothing
from sktime.forecasting.model_selection import temporal_train_test_split

y = load_airline()
y_train, y_test = temporal_train_test_split(y, test_size=36)
model = _ExponentialSmoothing(y_train, seasonal_periods=12, trend='add', seasonal='add', use_boxcox=True, initialization_method="estimated").fit()
model.fittedvalues
```

I can't figure out how to access the fitted values using sktime. Is it possible to get them?

**Comment by mloning** [comment, marked as answer]
Hi @eyalshafran you have two options:

Using the sktime interface for in-sample predictions (this should work for all forecasters):
```python
from sktime.datasets import load_airline
from sktime.forecasting.exp_smoothing import ExponentialSmoothing
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.forecasting.base import ForecastingHorizon

y = load_airline()
y_train, y_test = temporal_train_test_split(y, test_size=36)
model = ExponentialSmoothing(sp=12, trend='add', seasonal='add')
model.fit(y_train)
fh = ForecastingHorizon(y_train.index, is_relative=False)  # in-sample forecasting horizon
model.predict(fh)
```

Accessing the statsmodels used internally (this will only work for forecasters based on statsmodels): 
```python
model._fitted_forecaster.fittedvalues
```

**Reply by eyalshafran** [reply]
Thank you for the reply @mloning . I'm using your suggested code. I now face a different issue. Here is an example:
```
from sktime.datasets import load_airline
from sktime.forecasting.arima import ARIMA
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.model_selection import temporal_train_test_split

y = load_airline()
y_train, y_test = temporal_train_test_split(y, test_size=36)
model = ARIMA(order=(1, 1, 1), seasonal_order=(1, 1, 0, 12))
model.fit(y_train)
fh = ForecastingHorizon(y_train.index, is_relative=False)  # in-sample forecasting horizon
model.predict(fh)
```

This return the following error:
ValueError: In-sample predictions undefined for start=0 when d=1

In pmdarima the predict_in_sample method does not return an error. If I understand correctly this error is only relevant for the first point. Would it make more sense to return a NaN value for that point rather than returning an error for the whole forecast?

**Reply by mloning** [reply]
Hi @eyalshafran, yes, possibly. This is an error that comes from the `pmdarima` package as far as I understand. So perhaps it's worth asking there (there may be an issue on this already, open or closed) and link the discussion here?

**Reply by eyalshafran** [reply]
@mloning if I run the code in pdmarima I don't get an error message:
```
from pmdarima.arima import ARIMA as _ARIMA
from sktime.datasets import load_airline
from sktime.forecasting.model_selection import temporal_train_test_split

y = load_airline()
y_train, y_test = temporal_train_test_split(y, test_size=36)
model = _ARIMA(order=(1, 1, 1), seasonal_order=(1, 1, 0, 12))
model.fit(y_train)
model.predict_in_sample()
```
I believe the error is caused by calling the predict_in_sample with a start value smaller than d. This is how the pdmarima adapter calls the predict_in_sample. It is interesting that if you don't specify the start at all there is no error.

**Reply by mloning** [reply]
Alright, could you raise an issue about this with a reproducible example as above, once in sktime and once using pdmarima directly. I haven't fully understood yet where the error comes from, but would appreciate a PR if you think you it's related to sktime rather than pmdarima!

**Reply by eyalshafran** [reply]
I'll raise this issue with pdmarima. The error is a result of how predict_in_sample is called. There is inconsistency between calling it without any variables and calling it with start = 0 (which is what the pdmarima adapter does in sktime). Those two ways should be consistent.

**Reply by mloning** [reply]
Thanks @eyalshafran - that sounds like we need to adjust the `start` value that we pass to `predict_in_sample` depending on the forecasting horizon `fh` and model hyperparameter `d`. Would you be interested in opening a PR?
